### DISCUSSION TITLE
Is `request.data` truly immutable?

### DISCUSSION THREAD

**Discussion by roozmr** [encode/django-rest-framework#8867] [Potential Issue]
Hi, I am developing a POST API in which I send `multipart/form-data` content including an optional file field in request body.

Testing the API, I sent only text data in form fields and I encountered an error trying to add an extra field (like the code below) to `request.data` saying that `This QueryDict instance is immutable` and by Querydict it means `request.data`. 

```python
ticket = fetch_some_object()
request.data["ticket"] = ticket.id
```

**But** when I include files in form, API works without an error.

I traced the error and it seems when files are included in `request.data`, the `request`'s `self._mutable` is True while it is False when files are not included.

In the screen shot below from `django.http.request.QueryDict`, `AttributeError` is raised as I did not inlcude any files.

![image](https://user-images.githubusercontent.com/29716287/217008994-ea2adc0c-135c-4ecd-8a38-ba37802cc9d5.png)

In the next one raise statement is bypassed for `self._mutable` is True (attachments attribute includes `InMemoryUploadedFile`s).

![image](https://user-images.githubusercontent.com/29716287/217008909-1a7df91a-e728-4e5d-b008-ece160f0596d.png)

So, which one is the desired output?

**Comment by roozmr** [comment]
I found a previous closed issue #7633 addressing this problem. But I am using the latest DRF version and issue still persists.

**Reply by auvipy** [reply]
what about https://github.com/encode/django-rest-framework/pull/7634 ?

**Reply by roozmr** [reply]
I see, the commit is merged to master in 2020, yet the latest release has been out from Nov 2022 and I ran into the same error using the latest version.
